qiskit/algorithms/minimum_eigen_solvers/qaoa/qaoa_ansatz.py

```diff
@@ -0,0 +1,214 @@
+                as specified in the original paper.
+            initial_state: An optional initial state to use. If `None` is passed then a set of
+                Hadamard gates is applied as an initial state to all qubits.
+            mixer_operator: An optional custom mixer to use instead of the global X-rotations,
+                denoted as :math:`U(B, \beta)` in the original paper. Can be an operator or
+                an optionally parameterized quantum circuit.
+            name: A name of the circuit, default 'qaoa'
+        Raises:
+            TypeError: invalid input
+        """
+        super().__init__(name=name)
+        self._cost_operator = cost_operator
+        self._reps = reps
+        self._initial_state = initial_state
+        self._mixer_operator = mixer_operator
+
+        # set this circuit as a not-built circuit
+        self._num_qubits = None
+        self._num_parameters = 0
+        self._bounds = None
+        self._mixer = None
+        self._data = None
+
+    def _check_configuration(self, raise_on_failure: bool = True) -> bool:
+        valid = True
+
+        if self._cost_operator is None:
+            valid = False
+            if raise_on_failure:
+                raise AttributeError("The operator representing the cost of "
+                                     "the optimization problem is not set")
+        if self._reps is None or self._reps < 1:
+            valid = False
+            if raise_on_failure:
+                raise AttributeError("The integer parameter p, which determines the depth "
+                                     "of the circuit, either not set or set to non-positive value")
+
+        return valid
+
+    def _build(self) -> None:
+        """Build the circuit."""
+        if self._data:
+            return
+
+        self._check_configuration()
+        self._data = []
+
+        # calculate bounds, num_parameters, mixer
+        self._calculate_parameters()
+
+        # parametrize circuit and build it
+        param_vector = ParameterVector("θ", self._num_parameters)
+        circuit = self._construct_circuit(param_vector)
+
+        # append(replace) the circuit to this
+        self.add_register(circuit.num_qubits)
+        self.compose(circuit, inplace=True)
+
+    @property
+    def parameters(self) -> Set[Parameter]:
+        """Get the :class:`~qiskit.circuit.Parameter` objects in the circuit.
+
+        Returns:
+            A set containing the unbound circuit parameters.
+        """
+        self._build()
+        return super().parameters
+
+    def _calculate_parameters(self):
+        self._num_qubits = self._cost_operator.num_qubits
+
+        if isinstance(self._mixer_operator, QuantumCircuit):
+            self._num_parameters = (1 + self._mixer_operator.num_parameters) * self._reps
+            self._bounds = [(None, None)] * self._reps + [(None, None)] * \
+                           self._reps * self._mixer_operator.num_parameters
```

Comment:
Fix E127 continuation line over-indented for visual indent
```suggestion
            self._bounds = [(None, None)] * self._reps + [(None, None)] * \
                self._reps * self._mixer_operator.num_parameters
```